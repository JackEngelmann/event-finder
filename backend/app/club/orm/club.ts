import {
    Entity,
    Column,
    PrimaryGeneratedColumn,
    EntityRepository,
    Repository,
} from 'typeorm'

@Entity('club')
export class ClubDataModel {
    @Column({ type: 'text', nullable: true })
    address?: string | null

    @Column({ type: 'text', nullable: true })
    contact?: string | null

    @Column({
        type: 'text',
        nullable: true,
        charset: 'utf8mb4',
        collation: 'utf8mb4_general_ci',
    })
    description?: string | null

    @Column({ type: 'text', nullable: true })
    email?: string | null

    @PrimaryGeneratedColumn()
    id!: number

    // LEGACY
    @Column({ type: 'text', nullable: true })
    link?: string | null

    @Column()
    name!: string

    @Column({ type: 'text', nullable: true })
    region?: string | null

    @Column({
        type: 'text',
        nullable: true,
        charset: 'utf8mb4',
        collation: 'utf8mb4_general_ci',
    })
    specials?: string | null
}

@EntityRepository(ClubDataModel)
export class ClubRepository extends Repository<ClubDataModel> {
    async createAndSave(input: {
        address?: string
        contact?: string
        description?: string
        email?: string
        name: string
        region?: string
        specials?: string
    }) {
        const club = this.create(input)
        return await this.save(club)
    }
}
