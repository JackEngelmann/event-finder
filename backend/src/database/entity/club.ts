import { Entity, Column, PrimaryGeneratedColumn, Connection } from 'typeorm'
import { Logger } from '../../logger'

const logger = new Logger()

@Entity('club')
export class ClubDataModel {
    @Column({ type: 'text', nullable: true })
    address?: string | null

    @Column({ type: 'text', nullable: true })
    contact?: string | null

    @Column({ type: 'text', nullable: true })
    description?: string | null

    @Column({ type: 'text', nullable: true })
    email?: string | null

    @PrimaryGeneratedColumn()
    id!: number

    @Column({ type: 'text', nullable: true })
    imageUrl?: string | null

    @Column({ type: 'text', nullable: true })
    link?: string | null

    @Column()
    name!: string 

    @Column({ type: 'text', nullable: true })
    region?: string | null

    @Column({ type: 'text', nullable: true })
    specials?: string | null
}

export class ClubModel {
    private connection: Connection

    constructor(db: Connection) {
        this.connection = db
    }

    async createClub(input: {
        address?: string
        contact?: string
        description?: string
        email?: string
        imageUrl?: string
        link?: string
        name: string
        region?: string
        specials?: string
    }) {
        const club = new ClubDataModel()
        Object.assign(club, input)
        await this.connection.manager.save(club)
        return club.id
    }

    async updateClub(input: {
        address?: string
        contact?: string
        description?: string
        email?: string
        id: number
        imageUrl?: string
        link?: string
        name: string
        region?: string
        specials?: string
    }) {
        const club = await this.connection.manager.findOneOrFail<ClubDataModel>(
            'club',
            input.id
        )

        club.address = input.address || null
        club.contact = input.contact || null
        club.description = input.description || null
        club.email = input.email || null
        club.imageUrl = input.imageUrl || null
        club.link = input.link || null
        club.name = input.name
        club.region = input.region || null
        club.specials = input.specials || null

        await this.connection.manager.save(club)
    }

    async getClub(id: number) {
        const club = await this.connection.manager.findOne<ClubDataModel>('club', id)
        return club
    }

    async getClubs() {
        logger.info('club model: getClubs')
        const clubs = await this.connection.manager.find<ClubDataModel>('club')
        logger.info(JSON.stringify(clubs))
        return clubs
    }

    async deleteClub(id: number) {
        const club = await this.connection.manager.findOne<ClubDataModel>('club', id)
        await this.connection.manager.remove(club)
    }
}
