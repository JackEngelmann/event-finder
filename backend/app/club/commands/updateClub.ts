import { AppContext } from '../../infrastructure/appContext'
import { ClubRepository } from '../orm/club'
import { FileUpload } from 'graphql-upload'
import { LinkType } from '../../link/orm/link'
import { setLinksForClub } from '../../link/commands/setLinksForClub'
import { setImageUrlsForClub } from '../../image/commands/setImageUrlsForClub'
import logger from '../../infrastructure/logger'

export type UpdateClubInput = {
    address?: string
    contact?: string
    description?: string
    email?: string
    id: number
    image?: Promise<FileUpload>
    imageUrls?: string[]
    links?: LinkType[]
    name: string
    region?: string
    specials?: string
}

export async function updateClub(appContext: AppContext, input: UpdateClubInput) {
    const { db } = appContext
    const clubRepository = db.getCustomRepository(ClubRepository)
    try {
        await clubRepository.update(input.id, {
            address: input.address || null,
            contact: input.contact || null,
            description: input.description || null,
            email: input.email || null,
            id: input.id,
            name: input.name,
            region: input.region || null,
            specials: input.specials || null,
        })
        if (input.imageUrls) {
            await setImageUrlsForClub(appContext, {
                clubId: input.id,
                imageUrls: input.imageUrls,
            })
        }
        if (input.links) {
            await setLinksForClub(appContext, {
                clubId: input.id,
                links: input.links,
            })
        }
    } catch (error) {
        logger.error(error)
        throw error
    }
}
