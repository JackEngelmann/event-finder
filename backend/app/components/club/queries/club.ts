import { AppContext } from '../../../infrastructure/appContext'
import { ClubModel } from '../orm/club'

export function queryClub(appContext: AppContext, id: number) {
    const { db } = appContext
    const clubModel = new ClubModel(db)
    return clubModel.getClub(id)
}
