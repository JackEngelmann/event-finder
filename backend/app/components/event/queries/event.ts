import { AppContext } from '../../../infrastructure/appContext'
import { EventRepository } from '../orm/event'

export async function queryEvent(appContext: AppContext, id: number) {
    const { db } = appContext
    return await db.getCustomRepository(EventRepository).findOne({ id })
}
