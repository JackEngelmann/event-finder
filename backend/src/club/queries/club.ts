import { AppContext } from '../../infrastructure/appContext'
import { ClubRepository } from '../orm/club'

export async function queryClub(appContext: AppContext, id: number) {
    const { db } = appContext
    return await db.getCustomRepository(ClubRepository).findOne(id)
}
