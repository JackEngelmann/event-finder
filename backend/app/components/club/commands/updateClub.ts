import { AppContext } from '../../../infrastructure/appContext'
import { ClubModel } from '../orm/club'
import { FileUpload } from 'graphql-upload'
import { ClubImageModel } from '../../image/orm/clubImage'
import { LinkType } from '../../link/orm/link'
import { setLinksForClub } from '../../link/commands/setLinksForClub'

export type UpdateClubInput = {
    address?: string
    contact?: string
    description?: string
    email?: string
    id: number
    image?: Promise<FileUpload>
    imageUrls?: string[]
    links?: LinkType[]
    name: string
    region?: string
    specials?: string
}

export function updateClub(appContext: AppContext, input: UpdateClubInput) {
    const { db } = appContext
    const clubModel = new ClubModel(db)
    const clubImageModel = new ClubImageModel(db)
    return new Promise(async (resolve, reject) => {
        try {
            await clubModel.updateClub(input)
            await clubImageModel.setImageUrlsForClub(input.id, input.imageUrls)
            if (input.links) {
                await setLinksForClub(appContext, {
                    clubId: input.id,
                    links: input.links,
                })
            }
            resolve()
        } catch (err) {
            console.error(err)
            reject(err)
        }
    })
}
