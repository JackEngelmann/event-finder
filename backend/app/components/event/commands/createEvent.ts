import { AppContext } from '../../../infrastructure/appContext'
import { FileUpload } from 'graphql-upload'
import { createLinksForEvent } from '../../link/commands/createLinksForEvent'
import { LinkType } from '../../link/orm/link'
import { EventRepository } from '../orm/event'
import { setGenresForEvent } from '../../genre/commands/setGenresForEvent'
import { setImageUrlsForEvent } from '../../image/commands/setImageUrlsForEvent'
import logger from '../../../infrastructure/logger'

export type CreateEventInput = {
    admissionFee?: number
    admissionFeeWithDiscount?: number
    amountOfFloors?: number
    clubId: number
    date: string
    description?: string
    genreIds?: number[]
    image?: Promise<FileUpload>
    imageUrls?: string[]
    links?: LinkType[]
    minimumAge?: number
    name: string
    priceCategory?: number
    special?: string
}

export async function createEvent(appContext: AppContext, input: CreateEventInput) {
    const { db } = appContext
    try {
        const event = await db
            .getCustomRepository(EventRepository)
            .createAndSave(input)
        await setGenresForEvent(appContext, {
            eventId: event.id,
            genreIds: input.genreIds,
        })
        if (input.imageUrls) {
            await setImageUrlsForEvent(appContext, {
                eventId: event.id,
                imageUrls: input.imageUrls,
            })
        }
        if (input.links) {
            await createLinksForEvent(appContext, {
                eventId: event.id,
                links: input.links,
            })
        }
        return event.id
    } catch (error) {
        logger.error(error)
        throw error
    }
}
