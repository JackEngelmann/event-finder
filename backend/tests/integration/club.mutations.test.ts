import {
    clubFragment,
    createApolloTestServer,
    ApolloTestServer,
    ApolloHttpTestServer,
} from '../utils'
import { createEvent } from '../../src/event/commands/createEvent'
import { queryEvent } from '../../src/event/queries/event'
import {
    CreateClubInput,
    createClub,
} from '../../src/club/commands/createClub'
import { UpdateClubInput } from '../../src/club/commands/updateClub'
import { queryClub } from '../../src/club/queries/club'

const DB_NAME = 'clubmutationdb'

const createClubMutation = `
    mutation CreateClub($input: CreateClubInput!) {
        createClub(input: $input) {
            club { ${clubFragment} }
        }
    }
`
const updateClubMutation = `
    mutation UpdateClub($input: UpdateClubInput!) {
        updateClub(input: $input) {
            club { ${clubFragment} }
        }
    }
`
const deleteClubMutation = `
    mutation DeleteClub($id: Int!) {
        deleteClub(id: $id) {
            id
        }
    }
`
let apolloTestServer: ApolloTestServer | undefined
let apolloHttpTestServer: ApolloHttpTestServer | undefined

afterEach(async () => {
    if (apolloHttpTestServer) await apolloHttpTestServer.destroy()
    if (apolloTestServer) await apolloTestServer.destroy()
    apolloTestServer = undefined
    apolloHttpTestServer = undefined
})

describe('club mutations: ', () => {
    describe('create', () => {
        test('create club with only required information', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const input: CreateClubInput = {
                name: 'test',
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: createClubMutation,
                variables: { input },
            })

            // assert
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
        })

        test('create club with all information', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const input: CreateClubInput = {
                address: 'address',
                contact: 'contact',
                description: 'description',
                email: 'email',
                links: [{ href: 'test', type: 'FACEBOOK' }],
                name: 'test',
                region: 'region',
                specials: 'speicals',
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: createClubMutation,
                variables: { input },
            })

            // assert
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
        })

        test('create club without being admin should result in error', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                dbName: DB_NAME,
                isAdmin: false,
                insertTestData: true,
            })
            const input: CreateClubInput = {
                name: 'test',
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: createClubMutation,
                variables: { input },
            })

            // assert
            expect(result.data).toBeNull()
            expect(result.errors).toBeDefined()
            expect(result.errors).toMatchSnapshot()
        })
    })

    describe('update', () => {
        test('update club with only required information', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const input: UpdateClubInput = {
                name: 'updated-name',
                id: 1,
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: updateClubMutation,
                variables: { input },
            })

            // assert
            expect(result.errors).toBeUndefined()
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
        })

        test('update club with all information', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                insertTestData: true,
                dbName: DB_NAME,
            })
            const input: UpdateClubInput = {
                address: 'updated-address',
                contact: 'updated-contact',
                description: 'updated-description',
                email: 'updated-email',
                id: 1,
                links: [{ href: 'updated-link', type: 'FACEBOOK' }],
                name: 'updated-test',
                region: 'updated-region',
                specials: 'updated-speicals',
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: updateClubMutation,
                variables: { input },
            })

            // assert
            expect(result.errors).toBeUndefined()
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
        })

        test('update club without being admin should result in error', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                dbName: DB_NAME,
                isAdmin: false,
                insertTestData: true,
            })
            const input: UpdateClubInput = {
                name: 'test',
                id: 1,
            }

            // act
            const result = await apolloTestServer.client.mutate({
                mutation: updateClubMutation,
                variables: { input },
            })

            // assert
            expect(result.data).toBeNull()
            expect(result.errors).toBeDefined()
            expect(result.errors).toMatchSnapshot()
        })
    })

    describe('delete', () => {
        test('delete club', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const { client, appContext } = apolloTestServer
            const clubId = await createClub(appContext, {
                name: 'test',
            })
            const club = await queryClub(appContext, clubId)
            expect(club).toBeDefined()

            // act
            const result = await client.mutate({
                mutation: deleteClubMutation,
                variables: { id: clubId },
            })

            // assert
            expect(result.errors).toBeUndefined()
            const deletedClub = await queryClub(appContext, clubId)
            expect(deletedClub).toBeUndefined()
        })

        test('delete club without permission should yield error', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: false,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const { client } = apolloTestServer

            // act
            const result = await client.mutate({
                mutation: deleteClubMutation,
                variables: { id: 1 },
            })

            // assert
            expect(result.errors).toBeDefined()
            expect(result.data).toBeNull()
        })

        test('when deleting an event, the events the club did should be deleted too', async () => {
            // arrange
            apolloTestServer = await createApolloTestServer({
                isAdmin: true,
                dbName: DB_NAME,
                insertTestData: true,
            })
            const { client, appContext } = apolloTestServer
            const clubId = await createClub(appContext, {
                name: 'test',
            })
            const club = await queryClub(appContext, clubId)
            expect(club).toBeDefined()
            const connectedEventId = await createEvent(appContext, {
                clubId,
                name: 'connected event',
                date: 'some date',
            })
            const connectedEvent = await queryEvent(
                appContext,
                connectedEventId
            )
            expect(connectedEvent).toBeDefined()

            // act
            const result = await client.mutate({
                mutation: deleteClubMutation,
                variables: { id: clubId },
            })

            // assert
            expect(result.errors).toBeUndefined()
            const deletedClub = await queryClub(appContext, clubId)
            expect(deletedClub).toBeUndefined()
            const deletedEvent = await queryEvent(appContext, connectedEventId)
            expect(deletedEvent).toBeUndefined()
        })
    })
})
