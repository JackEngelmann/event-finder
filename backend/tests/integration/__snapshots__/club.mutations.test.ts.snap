// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`club mutations:  create create club with all information 1`] = `
Object {
  "data": Object {
    "createClub": Object {
      "club": Object {
        "address": "address",
        "contact": "contact",
        "description": "description",
        "email": "email",
        "id": 3,
        "imageUrls": Array [],
        "links": Array [
          Object {
            "href": "test",
            "type": "FACEBOOK",
          },
        ],
        "name": "test",
        "region": "region",
        "specials": "speicals",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`club mutations:  create create club with only required information 1`] = `
Object {
  "data": Object {
    "createClub": Object {
      "club": Object {
        "address": null,
        "contact": null,
        "description": null,
        "email": null,
        "id": 3,
        "imageUrls": Array [],
        "links": Array [],
        "name": "test",
        "region": null,
        "specials": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`club mutations:  create create club without being admin should result in error 1`] = `
Array [
  [GraphQLError: You need to be logged in as admin to mutate data],
]
`;

exports[`club mutations:  update update club with all information 1`] = `
Object {
  "data": Object {
    "updateClub": Object {
      "club": Object {
        "address": "updated-address",
        "contact": "updated-contact",
        "description": "updated-description",
        "email": "updated-email",
        "id": 1,
        "imageUrls": Array [],
        "links": Array [
          Object {
            "href": "updated-link",
            "type": "FACEBOOK",
          },
        ],
        "name": "updated-test",
        "region": "updated-region",
        "specials": "updated-speicals",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`club mutations:  update update club with only required information 1`] = `
Object {
  "data": Object {
    "updateClub": Object {
      "club": Object {
        "address": null,
        "contact": null,
        "description": null,
        "email": null,
        "id": 1,
        "imageUrls": Array [],
        "links": Array [],
        "name": "updated-name",
        "region": null,
        "specials": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`club mutations:  update update club without being admin should result in error 1`] = `
Array [
  [GraphQLError: You need to be logged in as admin to mutate data],
]
`;
